prov/efa/src/efa_av.c

```diff
@@ -317,7 +317,17 @@ int efa_conn_rdm_init(struct efa_av *av, struct efa_conn *conn)
 			return err;
 		}
 
-		ret = fi_av_insert(av->shm_rdm_av, smr_name, 1, &peer->shm_fiaddr, 0, NULL);
+		/*
+		 * The shm provider supports FI_AV_USER_ID flag. This flag
+		 * associates a user-assigned identifier with each av entry that is
+		 * returned with any completion entry in place of the AV's address.
+		 * EFA provider could use efa's fi_addr as the user id (passed as
+		 * as an input to the *fi_addr argument in fi_av_insert call)
+		 * With this change, there is no need to translate shm address back 
+		 * to efa address when reading shm cq entry.
+		 */
+		peer->shm_fiaddr = conn->fi_addr;
+		ret = fi_av_insert(av->shm_rdm_av, smr_name, 1, &peer->shm_fiaddr, FI_AV_USER_ID, NULL);
```

Comment:
Clarify - according to this statement
> EFA provider could use **efa's fi_addr** as the user id (passed as as an input to the *fi_addr argument in fi_av_insert call)

It sounds like we should use `&peer->efa_fiaddr` instead of `&peer->shm_fiaddr`. Could you explain?

Reply:
The `peer->shm_fiaddr` here is both an input and an output. When FI_AV_USER_ID flag is used, shm will record the passed in `peer->shm_fiaddr` (which was assigned as the conn->fi_addr) as user id. Then shm will overwrite `peer->shm_fiaddr` to the real shm addr. We cannot use `&peer->efa_fiaddr` in `fi_av_insert` call because it records the efa fi_addr and we do not want it to be overridden.

Reply:
Added comment in latest revision.